File: src/robot_life/behavior/manager.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from robot_life.event_engine.policy_layer import PolicyDecision


@dataclass(frozen=True)
class BehaviorPlan:
    target_behavior: str
    degraded_behavior: str | None
    required_resources: list[str]
    optional_resources: list[str]
    resume_previous: bool
    reason: str


class BehaviorManager:
    """Maps policy-approved scenes into concrete behavior plans."""

    def plan(
        self,
        scene: Any,
        *,
        rule: dict[str, Any],
        policy: PolicyDecision,
    ) -> BehaviorPlan:
        target_behavior = str(rule["target_behavior"])
        degraded_behavior = rule.get("degraded_behavior")
        required_resources = list(rule.get("required_resources", []))
        optional_resources = list(rule.get("optional_resources", []))
        resume_previous = bool(rule.get("resume_previous", True))
        return BehaviorPlan(
            target_behavior=target_behavior,
            degraded_behavior=degraded_behavior,
            required_resources=required_resources,
            optional_resources=optional_resources,
            resume_previous=resume_previous,
            reason=f"behavior_plan:{policy.response_level}:{getattr(scene, 'scene_type', 'unknown')}",
        )

File: Robot_Life_CPP/src/behavior/manager.cpp
#include "robot_life_cpp/behavior/manager.hpp"

namespace robot_life_cpp::behavior {

BehaviorPlan BehaviorManager::plan(
    const std::string& scene_type,
    const BehaviorRule& rule,
    const event_engine::PolicyDecision& policy) const {
  BehaviorPlan plan{};
  plan.target_behavior = rule.target_behavior;
  plan.degraded_behavior = rule.degraded_behavior;
  plan.required_resources = rule.required_resources;
  plan.optional_resources = rule.optional_resources;
  plan.resume_previous = rule.resume_previous;
  plan.reason = "behavior_plan:" + policy.response_level + ":" + scene_type;
  return plan;
}

}  // namespace robot_life_cpp::behavior
